driver-core/src/main/com/mongodb/internal/connection/OidcAuthenticator.java

```diff
@@ -0,0 +1,450 @@
+        }).finish(callback);
+    }
+
+    @Override
+    public void authenticate(final InternalConnection connection, final ConnectionDescription connectionDescription) {
+        assertFalse(connection.opened());
+        authenticationLoop(connection, connectionDescription);
+    }
+
+    @Override
+    void authenticateAsync(final InternalConnection connection, final ConnectionDescription connectionDescription,
+            final SingleResultCallback<Void> callback) {
+        beginAsync().thenRun(c -> {
+            assertFalse(connection.opened());
+            authenticationLoopAsync(connection, connectionDescription, c);
+        }).finish(callback);
+    }
+
+    private static boolean triggersRetry(@Nullable final Throwable t) {
+        if (t instanceof MongoSecurityException) {
+            MongoSecurityException e = (MongoSecurityException) t;
+            Throwable cause = e.getCause();
+            if (cause instanceof MongoCommandException) {
+                MongoCommandException commandCause = (MongoCommandException) cause;
+                return commandCause.getErrorCode() == 18;
+            }
+        }
+        return false;
+    }
+
+    private void authenticationLoop(final InternalConnection connection, final ConnectionDescription description) {
+        fallbackState = FallbackState.INITIAL;
+        while (true) {
+            try {
+                super.authenticate(connection, description);
+                break;
+            } catch (MongoSecurityException e) {
+                if (triggersRetry(e) && shouldRetryHandler()) {
+                    continue;
+                }
+                throw e;
+            }
+        }
+    }
+
+    private void authenticationLoopAsync(final InternalConnection connection, final ConnectionDescription description,
+            final SingleResultCallback<Void> callback) {
+        fallbackState = FallbackState.INITIAL;
+        beginAsync().thenRunRetryingWhile(
+                c -> super.authenticateAsync(connection, description, c),
+                e -> triggersRetry(e) && shouldRetryHandler()
+        ).finish(callback);
+    }
+
+    private byte[] evaluate(final byte[] challenge) {
+        if (isAutomaticAuthentication()) {
+            return prepareAwsTokenFromFileAsJwt();
+        }
+        byte[][] jwt = new byte[1][];
+        Locks.withLock(getMongoCredentialWithCache().getOidcLock(), () -> {
+            String cachedAccessToken = validatedCachedAccessToken();
+
+            if (cachedAccessToken != null) {
+                jwt[0] = prepareTokenAsJwt(cachedAccessToken);
+                fallbackState = FallbackState.PHASE_1_CACHED_TOKEN;
+            } else {
+                // cache is empty
+                OidcRequestCallback requestCallback = assertNotNull(getRequestCallback());
+                RequestCallbackResult result = requestCallback.onRequest(new OidcRequestContextImpl(CALLBACK_TIMEOUT));
```

Comment:
We should use interruptible locking in this PR where we may do blocking actions while holding a lock. See https://jira.mongodb.org/browse/JAVA-5140 and https://jira.mongodb.org/browse/JAVA-5189.

Reply:
I believe this is done, just the name needed to be updated.